// @remotion/studio/dist/components/Timeline/TimelineExpandedRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_row_layout_1 = require("./timeline-row-layout");
const TimelineEffectFieldRow_1 = require("./TimelineEffectFieldRow");
const TimelineEffectGroupRow_1 = require("./TimelineEffectGroupRow");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineFieldRow_1 = require("./TimelineFieldRow");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const TimelineExpandedRow = ({ node, depth, nestedDepth, getIsExpanded, toggleTrack, validatedLocation, nodePath, schema, }) => {
    const rowDepth = (0, timeline_row_layout_1.getExpandedRowDepth)({ nestedDepth, treeDepth: depth });
    if (node.kind === 'group') {
        if (node.effectInfo) {
            return (jsx_runtime_1.jsx(TimelineEffectGroupRow_1.TimelineEffectGroupRow, { label: node.label, nodePathInfo: node.nodePathInfo, effectIndex: node.effectInfo.effectIndex, effectSchema: node.effectInfo.effectSchema, documentationLink: node.effectInfo.documentationLink, nodePath: nodePath, validatedLocation: validatedLocation, rowDepth: rowDepth, getIsExpanded: getIsExpanded, toggleTrack: toggleTrack }));
        }
        const isExpanded = getIsExpanded(node.nodePathInfo);
        return (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(node.nodePathInfo), label: `${node.label} section`, disabled: false }), style: {
                height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
                paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
            }, children: jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label }) }));
    }
    if (node.field) {
        if (node.field.kind === 'effect-field') {
            return (jsx_runtime_1.jsx(TimelineEffectFieldRow_1.TimelineEffectFieldRow, { field: node.field, validatedLocation: validatedLocation, rowDepth: rowDepth, nodePath: nodePath }));
        }
        if (node.field.kind === 'sequence-field') {
            return (jsx_runtime_1.jsx(TimelineFieldRow_1.TimelineFieldRow, { field: node.field, validatedLocation: validatedLocation, rowDepth: rowDepth, nodePath: nodePath, schema: schema }));
        }
        throw new Error('Unexpected field kind: ' + JSON.stringify(node.field));
    }
    return (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {}), style: {
            height: (0, timeline_layout_1.getTreeRowHeight)(node),
            paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
        }, children: jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label }) }));
};
exports.TimelineExpandedRow = TimelineExpandedRow;
