// @remotion/studio/dist/components/Timeline/TimelineFieldRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineFieldRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const save_sequence_prop_1 = require("./save-sequence-prop");
const timeline_field_row_layout_1 = require("./timeline-field-row-layout");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const fieldRowBase = {
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const fieldName = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const Value = ({ field, nodePath, validatedLocation, schema, codeValue }) => {
    const { getDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { setDragOverrides, clearDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        var _a;
        return nodePath === null
            ? undefined
            : ((_a = getDragOverrides(nodePath)) !== null && _a !== void 0 ? _a : {})[field.key];
    }, [getDragOverrides, nodePath, field.key]);
    const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
        codeValue,
        dragOverrideValue,
        defaultValue: field.fieldSchema.default,
        shouldResortToDefaultValueIfUndefined: true,
    });
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const clientId = previewServerState.type === 'connected'
        ? previewServerState.clientId
        : null;
    const onSave = (0, react_1.useCallback)((value) => {
        if (!codeValue || !codeValue.canUpdate) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!clientId) {
            return Promise.reject(new Error('Not connected to studio server'));
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        const stringifiedValue = JSON.stringify(value);
        if (value === codeValue.codeValue) {
            return Promise.resolve();
        }
        if (defaultValue === stringifiedValue &&
            codeValue.codeValue === undefined) {
            return Promise.resolve();
        }
        return (0, save_sequence_prop_1.saveSequenceProp)({
            fileName: validatedLocation.source,
            nodePath,
            fieldKey: field.key,
            value,
            defaultValue,
            schema,
            setCodeValues,
            clientId,
        });
    }, [
        codeValue,
        clientId,
        field.fieldSchema.default,
        field.key,
        nodePath,
        schema,
        setCodeValues,
        validatedLocation,
    ]);
    const onDragValueChange = (0, react_1.useCallback)((value) => {
        if (nodePath === null) {
            throw new Error('Cannot drag value');
        }
        setDragOverrides(nodePath, field.key, value);
    }, [setDragOverrides, nodePath, field.key]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        if (nodePath === null) {
            throw new Error('Cannot clear drag value');
        }
        clearDragOverrides(nodePath);
    }, [clearDragOverrides, nodePath]);
    return (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: codeValue, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue }));
};
const TimelineFieldRow = ({ field, validatedLocation, rowDepth, nodePath, schema }) => {
    var _a, _b;
    const { codeValues: visualModeCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const codeValuesForOverride = remotion_1.Internals.getCodeValuesCtx(visualModeCodeValues, nodePath);
    const codeValue = (_a = codeValuesForOverride === null || codeValuesForOverride === void 0 ? void 0 : codeValuesForOverride[field.key]) !== null && _a !== void 0 ? _a : null;
    const style = (0, react_1.useMemo)(() => {
        return {
            ...fieldRowBase,
            height: field.rowHeight,
        };
    }, [field.rowHeight]);
    const labelRowStyle = (0, react_1.useMemo)(() => (0, timeline_field_row_layout_1.getTimelineFieldLabelRowStyle)(rowDepth), [rowDepth]);
    if (codeValue === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {}), style: style, children: [
            jsx_runtime_1.jsx("div", { style: labelRowStyle, children: jsx_runtime_1.jsx("span", { style: fieldName, children: (_b = field.description) !== null && _b !== void 0 ? _b : field.key }) }), codeValue.canUpdate ? (jsx_runtime_1.jsx(Value, { field: field, nodePath: nodePath, validatedLocation: validatedLocation, schema: schema, codeValue: codeValue })) : (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineNonEditableStatus, { propStatus: codeValue }))] }));
};
exports.TimelineFieldRow = TimelineFieldRow;
