// @remotion/studio/dist/components/Timeline/TimelineExpandedTrackKeyframes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedTrackKeyframes = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const get_left_of_timeline_slider_1 = require("../../helpers/get-left-of-timeline-slider");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
const row = {
    position: 'relative',
};
const separator = {
    height: 1,
    backgroundColor: colors_1.TIMELINE_TRACK_SEPARATOR,
};
const section = {
    borderBottom: `1px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const diamondBase = {
    position: 'absolute',
    width: 8,
    height: 8,
    backgroundColor: colors_1.LIGHT_TEXT,
    borderRadius: 1,
    boxShadow: '0 0 0 1px rgba(0, 0, 0, 0.4)',
    pointerEvents: 'none',
};
const getNodeKeyframes = ({ node, nodePath, codeValues, }) => {
    var _a, _b;
    if (node.kind !== 'field' || node.field === null) {
        return [];
    }
    if (node.field.kind === 'sequence-field') {
        return (0, get_timeline_keyframes_1.getTimelineKeyframes)((_a = remotion_1.Internals.getCodeValuesCtx(codeValues, nodePath)) === null || _a === void 0 ? void 0 : _a[node.field.key]);
    }
    const effectStatus = remotion_1.Internals.getEffectCodeValuesCtx({
        codeValues,
        nodePath,
        effectIndex: node.field.effectIndex,
    });
    return (0, get_timeline_keyframes_1.getTimelineKeyframes)(effectStatus.type === 'can-update-effect'
        ? (_b = effectStatus.props) === null || _b === void 0 ? void 0 : _b[node.field.key]
        : null);
};
const TimelineExpandedTrackKeyframesInner = ({ nodePathInfo, sequence }) => {
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const timelineWidth = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { codeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)({
        sequence,
        nodePathInfo,
        getDragOverrides: () => ({}),
        codeValues,
    }), [codeValues, nodePathInfo, sequence]);
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)({ nodes: tree, getIsExpanded }), [tree, getIsExpanded]);
    const expandedHeight = (0, react_1.useMemo)(() => (0, timeline_layout_1.getExpandedTrackHeight)({
        sequence,
        nodePathInfo,
        getIsExpanded,
        codeValues,
    }), [codeValues, getIsExpanded, nodePathInfo, sequence]);
    const rows = (0, react_1.useMemo)(() => flat.map(({ node }) => ({
        height: (0, timeline_layout_1.getTreeRowHeight)(node),
        keyframes: getNodeKeyframes({
            node,
            nodePath: nodePathInfo.sequenceSubscriptionKey,
            codeValues,
        }),
        key: JSON.stringify(node.nodePathInfo),
    })), [codeValues, flat, nodePathInfo.sequenceSubscriptionKey]);
    return (jsx_runtime_1.jsx("div", { style: {
            height: expandedHeight + timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM,
        }, children: jsx_runtime_1.jsx("div", { style: { ...section, height: expandedHeight }, children: rows.map(({ height, keyframes, key }, i) => {
                return (jsx_runtime_1.jsxs(react_1.default.Fragment, { children: [i > 0 ? jsx_runtime_1.jsx("div", { style: separator }) : null, jsx_runtime_1.jsx("div", { style: { ...row, height }, children: timelineWidth === null
                                ? null
                                : keyframes.map((keyframe) => (jsx_runtime_1.jsx("div", { style: {
                                        ...diamondBase,
                                        left: (0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(keyframe.frame, videoConfig.durationInFrames, timelineWidth) - timeline_layout_1.TIMELINE_PADDING,
                                        top: height / 2,
                                        transform: 'translate(-50%, -50%) rotate(45deg)',
                                    }, title: `Keyframe at frame ${keyframe.frame}` }, String(keyframe.frame)))) })
                    ] }, key));
            }) }) }));
};
exports.TimelineExpandedTrackKeyframes = react_1.default.memo(TimelineExpandedTrackKeyframesInner);
